starfish/imagestack/physical_coordinate_calculator.py

```diff
@@ -0,0 +1,173 @@
+
+import xarray as xr
+
+from starfish.imagestack import indexing_utils
+from starfish.types import (
+    Coordinates,
+    Indices,
+    PHYSICAL_COORDINATE_DIMENSION,
+    PhysicalCoordinateTypes
+)
+
+
+def calc_new_physical_coords_array(physical_coordinates: xr.DataArray,
+                                   stack_shape: Mapping[Indices, int],
+                                   indexers) -> xr.DataArray:
+    """Calculates the resulting coordinates array from indexing on each dimension in indexers
+
+    Parameters
+    ----------
+    physical_coordinates: xarray
+        xarray that holds the min/max values of physical coordinates per tile (Round, Ch, Z).
+    stack_shape : Dict[str, int]
+        The shape of the image tensor by categorical index (channels, imaging rounds, z-layers)
+    indexers : Dict[str, (int/slice)]
+        A dictionary of dim:index where index is the value or range to index the dimension
+    """
+    new_coords = physical_coordinates.copy()
+    # index by R, CH, V
+    key = {Indices.ROUND.value: indexers[Indices.ROUND.value],
+           Indices.CH.value: indexers[Indices.CH.value],
+           Indices.Z.value: indexers[Indices.Z.value]}
+    new_coords = indexing_utils.index_keep_dimensions(new_coords, key)
+    # check if X or Y dimension indexed, if so recalculate physcal coordinate min/max values
+    if _needs_coords_recalculating(indexers[Indices.X.value], indexers[Indices.Y.value]):
+        _recalculate_physical_coordinate_ranges(stack_shape, indexers, new_coords)
+    return new_coords
+
+
+def _needs_coords_recalculating(x_indexers, y_indexers) -> bool:
+    if isinstance(x_indexers, int) or isinstance(y_indexers, int):
+        return True
+    return not (x_indexers.start is x_indexers.stop is y_indexers.start is y_indexers.stop is None)
+
+
+def _recalculate_physical_coordinate_ranges(stack_shape: Mapping[Indices, int],
+                                            indexers: Mapping[str, Union[int, slice]],
+                                            coords_array: xr.DataArray
+                                            ) -> None:
+    """Iterates through coordinates array and recalculates x, y min/max physical coordinate values
+     based on how the x and y dimensions were indexed
+
+    Parameters
+    ----------
+    indexers : Dict[str, (int/slice)]
+        A dictionary of dim:index where index is the value or range to index the dimension
+
+    coords_array: xarray
+        xarray that holds the min/max values of physical coordinates per tile (Round, Ch, Z).
+
+    """
+    for _round in range(coords_array.sizes[Indices.ROUND]):
+        for ch in range(coords_array.sizes[Indices.CH]):
+            for z in range(coords_array.sizes[Indices.Z]):
+                indices = {
+                    Indices.ROUND: _round,
+                    Indices.CH: ch,
+                    Indices.Z: z
+                }
+                xmin, xmax = _recalculate_physical_coordinate_range(
```

Comment:
I have a preference for:

```suggestion
                xmin, xmax, ymin, ymax = coords_array[indices].loc[
                    dict(physical_coordinate=[
                        PhysicalCoordinateTypes.X_MIN,
                        PhysicalCoordinateTypes.X_MAX,
                        PhysicalCoordinateTypes.Y_MIN,
                        PhysicalCoordinateTypes.Y_MAX
                    ])]
                xmin, xmax = _recalculate_physical_coordinate_range(
                    xmin, xmax,
                    stack_shape[Indices.X.value],
                    indexers[Indices.X.value])
                ymin, ymax = _recalculate_physical_coordinate_range(
                    ymin, ymax,
                    stack_shape[Indices.Y.value],
                    indexers[Indices.Y.value])
                coords_array[indices].loc[
                    dict(physical_coordinate=[
                        PhysicalCoordinateTypes.X_MIN,
                        PhysicalCoordinateTypes.X_MAX,
                        PhysicalCoordinateTypes.Y_MIN,
                        PhysicalCoordinateTypes.Y_MAX
                    ])] = [xmin, xmax, ymin, ymax]
```

And `_recalculate_physical_coordinate_range` would accept the following arguments:
1. physical_coordinate_start
2. physical_coordinate_end
3. pixel_size
4. pixel_slice

Rationale: when I do cropping _before_ `ImageStack` is loaded, I probably won't have `coords_array`.  If you structure `_recalculate_physical_coordinate_range` this way, I can use it without modification.  Probably... :)

Reply:
OH this is a really good call.